xbmc/ApplicationPlayer.cpp

```diff
@@ -116,6 +116,15 @@ bool CApplicationPlayer::OpenFile(const CFileItem& item, const CPlayerOptions& o
     }
   }
 
+  //check if player has changed?
+  if (player && (player->m_name != newPlayer))
```

Comment:
why do we have a player here? old one should have been released in ClosePlayer

Reply:
A few lines above the Current player will be retrieved with GetInternal() then is a check if player != null and if it's playing...  Some code. So in most cases it isn't playing and the if isn't executed. New player then isn't created because it's not null. 

My code just releases it when the player name changed. 

I'll think the code is not straight forward in some cases. 

The m_player has to be null that a new player is created. 

But which function closes it?

Reply:
I have tracked it down that m_player is definitely not released. In openFile on the GetInternal call a player is returned and worked with.  That's the reason why not the correct player is created and assigned to m_player.

Reply:
I still don't understand what's exactly going on here. If I start a video and stop it, the player is released on ClosePlayer. Next time I open a file, there is no player instance. Is this different on your system? Why?

Reply:
Please try the following:

 * setup PCF with my xml from the ticket
 * create a bat file under c:\temp\test.bat with the content:

@echo off
echo Playing "%1"
pause

 * Ensure that you have a movie in your DB with "4K" in the filename (without quotes)
 * start latest kodi nightly
 * play a movie (which NOT has 4K in it) ->Video Player should open
 * press X to stop it
 * play another movie (which NOT has 4K in it) ->Video Player should open
 * press X to stop it
 * play a movie which HAS 4K in its filename -> test.bat should open
 * until here all is ok
 * play the first movie  (which NOT has 4K in it) ->Video Player should open
 * But sadly until now EVERY movie is opened in the external player (batch file in our case)
   Thats the issue!

I have debugged it and found out that "player", at the lines where i have commited the new code, has a reference to the last player (in our case MPC-HC or our batch file). This reference comes from the function GetInternal() which returns the global m_Player variable.
I'll think the problem is that "player" will be filled with the retuirn value of GetInternal and worked with but not released.

Just when "player" is null, a new player is created.

My fix now checks the name of the player and releases it if it differs from the new/reqeuested one.

pOpY

Reply:
If this happens we have another issue to fix. When a video is stopped, ClosePlayer should be called. This seems not to be the case for external player, otherwise player would be released.

Reply:
You are right, have looked at the code in the Application.cpp at:

void CApplication::StopPlaying()
{
  int iWin = CServiceBroker::GetGUI()->GetWindowManager().GetActiveWindow();
  if (m_appPlayer.IsPlaying())
  {
    m_appPlayer.ClosePlayer();

So when is StopPlaying() called, i assume when the inputmanager gets an stop key.

I'll think the issue could be that:

 * External Player will be launched -> seperate thread starts and executes (Process function) until the external application is stopped
 * Any other application (Harmony Remote, Yatse) is closing the external app WITHOUT sending stop to kodi
 * kodi doesn't call the StopPlaying() so the player isnt released.

**I am not so deep in the kodi kodi than you.**
**Can you confirm that?**

My idea is that the stopping external player thread threaf is releasing the ApplicationPlayer (=m_pPlayer variable) or calling StopPlaying().

Reply:
what do you think of adding ClosePlayer() to the end of the Process function of ExternalPlayer.cpp:

>  // When external player is closed NOT with kodi stop function ClosePlayer() doesn't get called > and the player isn't released
>   g_application.GetAppPlayer().ClosePlayer();

>  if (!ret || (m_playOneStackItem && g_application.CurrentFileItem().IsStack()))
>    m_callback.OnPlayBackStopped();
>  else
>    m_callback.OnPlayBackEnded();
>}

Reply:
@Rechi spotted the root cause: https://github.com/xbmc/xbmc/pull/12353#pullrequestreview-145583868

Reply:
Nice find @Rechi
Tested and working here also.
Changed PR will come soon...